src/LinAlg/matrix.h

```diff
@@ -753,6 +755,32 @@ namespace utl //! General utility classes and functions.
       n[2] = n_3;
     }
 
+    //! \brief Constructor fill matrix from istream.
+    matrix3d(const std::string& fname)
+    {
+      std::ifstream is(fname, std::ios::in);
+      for (size_t i=0; i<3; ++i)
+        is >> n[i];
+      double num;
+      elem.clear();
+      while (is >> num)
+        elem.push_back(num);
+      if (n[0]*n[1]*n[2] != elem.size())
+        std::cerr << "matrix3d: error reading from file, expected size "
+                  << n[0]*n[1]*n[2] << " but got " << elem.size() << std::endl;
+    }
+    //! \brief Dump contents to output stream.
+    void dump(const std::string& fname, size_t precision = 16)
+    {
+      std::ofstream os(fname);
+      os << std::setprecision(precision);
+      os << n[0] << std::endl << n[1] << std::endl << n[2] << std::endl;
+      for (size_t k=1; k<=n[2]; ++k)
+        for (size_t j=1; j<=n[1]; ++j)
+          for (size_t i=1; i<=n[0]; ++i)
+            os << this->operator()(i,j,k) << "\n";
+    }
+
     //! \brief Resize the matrix to dimension \f$n_1 \times n_2 \times n_3\f$.
```

Comment:
Well, I would rather not have file operations inside low-level classes like this. Instead the constructor and dump methods above should have an std::istream& and std::ostream& argument instead, such that it could be decided outside if these are from/to files or other streams. Or even better, implement as operator<< and operator>> instead (like for the vector and matrix classes in this file).

Moreover, std::endl is the same as "\n" followed by a flush, so in line 777 it would be sufficient with three "\n" instead. No point in flushing except for at the end of the print.